### ISSUE TITLE
The behavior of "vagrant box add -f" is confusing with libvirt

### ISSUE THREAD

**Issue by xie-jinwei** [vagrant-libvirt/vagrant-libvirt#1351]
### Steps to reproduce
1. Use packer to build a vagrant libvirt box called A, and output it to ./A/package.box
2. Run: `vagrant box add -f --name A ./A/package.box`
3. Write a Vagrantfile using box A
4. Run `vagrant up` with the new Vagrantfile
5. Run `vagrant destroy`
6. Use packer to rebuild the vagrnat libvirt box called A (with some updates in the box), and output it to ./A1/package.box
7. Run: `vagrant box add -f --name A ./A1/package.box`
8. Run `vagrant up`

### Expected behaviour
We expect that the box image in libvirt VM to be updated with the new box.

### Actual behaviour
We found that the box image is not updated.

We actually have some theories for why this is happening: because `vagrant up --provider libvirt` will only upload base box image to libvirt when volume `${BASE_BOX_NAME}_vagrant_image_0_box.img` is not found in libvirt, so after a force box update using `vagrant add -f`, another `vagrant up` will not upload the base box image to libvirt again, causing the box image to not be updated.

### System configuration
**OS/Distro version:**: lsb_release -a
**Libvirt version:** 6.0.0
**Output of `vagrant version; vagrant plugin list`:**
Installed Version: 2.2.18
Latest Version: 2.2.18
 
You're running an up-to-date version of Vagrant!
vagrant-env (0.0.3, global)
  - Version Constraint: > 0
vagrant-libvirt (0.5.3, global)
vagrant-mutate (1.2.0, global)
vagrant-proxyconf (2.0.10, global)
  - Version Constraint: > 0
vagrant-vbguest (0.30.0, global)
  - Version Constraint: > 0

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**
```
Delete this text and drag and drop the log file for github to attach and add a link here
```

**A Vagrantfile to reproduce the issue:**
```
We are using a local built box, so we don't think this part is applicable
```

**Are you using upstream vagrant package or your distros package?**
Upstream / Distro

**Comment by electrofelix** [comment]
If you set the box version each time it will work as the version is embedded in the volume name, without it you get the default value of `''`:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/d1bad669b6e6a1c7fbddb29b0de02990d7158459/lib/vagrant-libvirt/action/handle_box_image.rb#L125-L134

At the very least we're missing some kind of warning at the following point to let end-users know that without the version the image won't be updated and at least then it wouldn't cause this surprise:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/d1bad669b6e6a1c7fbddb29b0de02990d7158459/lib/vagrant-libvirt/action/handle_box_image.rb#L130-L132

With a quick look, I see one fix would be to compare the `actual-size` values returned by `qemu-img info <box-disk>` against the volume information returned from libvirt and replace if different. However I'm not sure that is safe, as what happens to existing machines using the old image. So I'd probably go with embedding the actual size returned by `qemu-img info <box-disk>` into the volume name to be uploaded to as the fallback when no box version is available, and maybe still emit a warning that there is no guarantee that you can't accidentally end up with a duplicate box. I'm wary about trying to use the timestamp as that appears to be the timestamp of when the box is added to your vagrant directory rather than the time the box was built as there doesn't appear to be anything that captures that (if vagrant's box unpacking preserved timestamps it would have worked but it doesn't appear to me that is the case).

Would embedding the size work in your case or do you think just using the timestamp and warning that it can vary from env to env and it would be better to set an explicit box version is better?

**Comment by xie-jinwei** [comment]
Thanks for your reply, for my case, I think a warning message should be enough (altough it could be confusing with `vagrant add -f`)

Also, I don't know how to assign a version number to our local built libvirt boxes, I found this on [Packer Vagrant builder](https://www.packer.io/docs/builders/vagrant), but it seems that this only set the base box version, not the output box version
![image](https://user-images.githubusercontent.com/86954937/133708947-bf6ad2aa-f798-4a07-a1e0-c7768ad483b5.png)

**Comment by electrofelix** [comment]
I forgot the version isn't embedded, so the current workaround would be to generate a catalog metadata json file locally that references the box and point the `vagrant box add ...` command to use that instead of the box directly. https://stackoverflow.com/a/32037117

It should be possible to use the shell post processor in packer to regenerate the json file contents needed on each packer run and use a timestamp in that case as a simple version until I can get back to updating the box handling and tests to check the size correctly.

**Comment by xie-jinwei** [comment]
> I forgot the version isn't embedded, so the current workaround would be to generate a catalog metadata json file locally that references the box and point the `vagrant box add ...` command to use that instead of the box directly. https://stackoverflow.com/a/32037117
> 
> It should be possible to use the shell post processor in packer to regenerate the json file contents needed on each packer run and use a timestamp in that case as a simple version until I can get back to updating the box handling and tests to check the size correctly.

OK, thank you, my current work around is just to delete the box image volume in libvirt everytime before running `vagrant box add`, but I think your method is much more decent, I will use it for our project.

**Comment by electrofelix** [comment]
I thought this was fixed, but it looks like I made an incorrect assumption on the default box version when it's not specified, in that it is set to 0 rather than empty when not specified. Don't ask me how I missed this as it appears it's been this way since about 2014.

**Comment by electrofelix** [comment]
Should now be fixed with #1512 and should therefore be properly fixed with the next release
